streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -177,6 +126,137 @@ void openDB(final ProcessorContext context) {
+            cache = new LRUCache(BLOCK_CACHE_SIZE);
+            tableConfig.setBlockCache(cache);
+            tableConfig.setBlockSize(BLOCK_SIZE);
+
+            filter = new BloomFilter();
+            tableConfig.setFilter(filter);
+
+            dbOptions = new DBOptions();
+            dbOptions.setCreateIfMissing(true);
+            dbOptions.setErrorIfExists(false);
+            dbOptions.setInfoLogLevel(InfoLogLevel.ERROR_LEVEL);
+            // this is the recommended way to increase parallelism in RocksDb
+            // note that the current implementation of setIncreaseParallelism affects the number
+            // of compaction threads but not flush threads (the latter remains one). Also
+            // the parallelism value needs to be at least two because of the code in
+            // https://github.com/facebook/rocksdb/blob/62ad0a9b19f0be4cefa70b6b32876e764b7f3c11/util/options.cc#L580
+            // subtracts one from the value passed to determine the number of compaction threads
+            // (this could be a bug in the RocksDB code and their devs have been contacted).
+            dbOptions.setIncreaseParallelism(Math.max(Runtime.getRuntime().availableProcessors(), 2));
+
+            columnFamilyOptions = new ColumnFamilyOptions();
+            columnFamilyOptions.optimizeFiltersForHits();
+            columnFamilyOptions.setTableFormatConfig(tableConfig);
+            columnFamilyOptions.setWriteBufferSize(WRITE_BUFFER_SIZE);
+            columnFamilyOptions.setCompressionType(COMPRESSION_TYPE);
+            columnFamilyOptions.setCompactionStyle(COMPACTION_STYLE);
+            columnFamilyOptions.setMaxWriteBufferNumber(MAX_WRITE_BUFFERS);
+
+            wOptions = new WriteOptions();
+            wOptions.setDisableWAL(true);
+
+            fOptions = new FlushOptions();
+            fOptions.setWaitForFlush(true);
+
+            final Map<String, Object> configs = context.appConfigs();
+            final Class<RocksDBConfigSetter> configSetterClass =
+                    (Class<RocksDBConfigSetter>) configs.get(StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG);
+
+            final RocksDBConfigSetter configSetter;
+            if (configSetterClass != null) {
+                configSetter = Utils.newInstance(configSetterClass);
+                configSetter.setConfig(name, dbOptions, columnFamilyOptions, configs);
+            } else {
+                configSetter = null;
+            }
+
+            if (prepareForBulkload) {
+                dbOptions.setMaxBackgroundFlushes(4);
+                columnFamilyOptions.setDisableAutoCompactions(true);
+                columnFamilyOptions.setLevel0FileNumCompactionTrigger(1 << 30);
+                columnFamilyOptions.setLevel0SlowdownWritesTrigger(1 << 30);
+                columnFamilyOptions.setLevel0StopWritesTrigger(1 << 30);
+            }
+
+            if (configSetter != null) {
+                configSetter.close(name, dbOptions, columnFamilyOptions);
```

Comment:
configSetter.close() will clean up any resources constructed in configSetter.setConfig, and should only be called in RocksDBStore#close()